quickwit-common/src/uri.rs

```diff
@@ -83,48 +90,94 @@ impl Uri {
 
                 path.replace_range(0..1, &home_dir_path);
             }
-
-            if !path.starts_with('/') {
+            if Path::new(&path).is_relative() {
                 let current_dir = env::current_dir().context(
                     "Failed to resolve current working directory: dir does not exist or \
                      insufficient permissions.",
                 )?;
                 path = current_dir.join(path).to_string_lossy().to_string();
             }
-
             path = normalize_path(Path::new(&path))
                 .to_string_lossy()
                 .to_string();
         }
-
         Ok(Self {
             uri: format!("{}{}{}", protocol, PROTOCOL_SEPARATOR, path),
             protocol_idx: protocol.len(),
         })
     }
 
-    /// Returns the URI extension.
+    /// Constructs a [`Uri`] from a properly formatted string `<protocol>://<path>` where `path` is
+    /// normalized. Use this method exclusively for trusted input.
+    pub fn new(uri: String) -> Self {
+        let protocol_idx = uri
+            .find(PROTOCOL_SEPARATOR)
+            .expect("URI lacks protocol separator. Use `Uri::new` exclusively for trusted input.");
+        Self { uri, protocol_idx }
+    }
+
+    /// Returns the extension of the URI.
     pub fn extension(&self) -> Option<Extension> {
         Path::new(&self.uri)
             .extension()
             .and_then(OsStr::to_str)
             .and_then(Extension::maybe_new)
     }
 
-    /// Returns the uri protocol.
+    /// Returns the protocol of the URI.
     pub fn protocol(&self) -> &str {
         &self.uri[..self.protocol_idx]
     }
 
-    /// Returns the file path from the uri.
-    /// Useful only for `file://` protocol Uri.
+    /// Returns the file path of the URI.
+    /// Applies only to `file://` URIs.
     pub fn filepath(&self) -> Option<&Path> {
-        if self.protocol() == "file" {
+        if self.protocol() == FILE_PROTOCOL {
             self.uri.strip_prefix("file://").map(Path::new)
         } else {
             None
         }
     }
+
+    /// Consumes the [`Uri`] struct and returns the normalized URI as a string.
+    pub fn into_string(self) -> String {
+        self.uri
+    }
+
+    /// Creates a new [`Uri`] with `path` adjoined to `self`.
+    /// Fails if `path` is absolute, it replaces the current path.
+    pub fn join(&self, path: &str) -> anyhow::Result<Self> {
+        if Path::new(path).is_absolute() {
+            bail!(
+                "Cannot join URI `{}` with absolute path `{}`.",
+                self.uri,
+                path
+            );
+        }
+        let joined = match self.protocol() {
+            FILE_PROTOCOL => Path::new(&self.uri)
+                .join(path)
+                .to_string_lossy()
+                .to_string(),
+            RAM_PROTOCOL | S3_PROTOCOL => format!(
```

Comment:
I would just do this for any other protocol and specially treat `file://` like you have above.
The semantic being that when you join a path, you need it with a separator. we also have `postgres` `postgresql` and many more.

Reply:
I agree, adding `as_str`.

`join` does not really make sense for `postgres` URIs that is why I wanted those to fall in the catchall category and raise. However, the way I've implemented this function prevents the use of protocols that we have not anticipated, which is bound to break at some point. So I'm going to go with your recommendation for that reason :)